modules/aws/test/src/smithy4s/aws/query/AwsQueryCodecSpec.scala

```diff
@@ -0,0 +1,350 @@
+    implicit val schema: Schema[ByteArray] = bytes
+    val expected = "Zm9vYmFy"
+    checkContent(expected, ByteArray("foobar".getBytes()))
+  }
+
+  test("struct") {
+    case class Foo(x: String, y: Option[String])
+    object Foo {
+      implicit val schema: Schema[Foo] = {
+        val x = string.required[Foo]("x", _.x)
+        val y = string.optional[Foo]("y", _.y)
+        struct(x, y)(Foo.apply)
+      }
+    }
+
+    val expected = "x=value-x&y=value-y"
+
+    checkContent(expected, Foo("value-x", Some("value-y")))
+  }
+
+  test("struct: empty optional") {
+    case class Foo(x: String, y: Option[String])
+    object Foo {
+      implicit val schema: Schema[Foo] = {
+        val x = string.required[Foo]("x", _.x)
+        val y = string.optional[Foo]("y", _.y)
+        struct(x, y)(Foo.apply)
+      }
+    }
+
+    val expected = "x=value-x"
+
+    checkContent(expected, Foo("value-x", None))
+  }
+
+  test("struct: custom names") {
+    case class Foo(x: String, y: Option[String])
+    object Foo {
+      implicit val schema: Schema[Foo] = {
+        val x = string.required[Foo]("x", _.x).addHints(XmlName("xx"))
+        val y = string.optional[Foo]("y", _.y).addHints(XmlName("y:y"))
+        struct(x, y)(Foo.apply)
+      }
+    }
+
+    val expected = "xx=value-x&y%3Ay=value-y"
+
+    checkContent(expected, Foo("value-x", Some("value-y")))
+  }
+
+  test("list") {
+    case class Foo(foos: List[Int])
+    object Foo {
+      implicit val schema: Schema[Foo] = {
+        val foos = list(int).required[Foo]("foos", _.foos)
+        struct(foos)(Foo.apply)
+      }
+    }
+
+    val expected = "foos.member.1=1&foos.member.2=2&foos.member.3=3"
+
+    checkContent(expected, Foo(List(1, 2, 3)))
+  }
+
+  test("list: custom names") {
+    case class Foo(foos: List[Int])
+    object Foo {
+      implicit val schema: Schema[Foo] = {
+        val foos = list(int.addHints(XmlName("x")))
+          .required[Foo]("foos", _.foos)
+        struct(foos)(Foo.apply)
+      }
+    }
+
+    val expected = "foos.x.1=1&foos.x.2=2&foos.x.3=3"
+
+    checkContent(expected, Foo(List(1, 2, 3)))
+  }
+
+  test("list: flattened") {
+    case class Foo(foos: List[Int])
+    object Foo {
+      implicit val schema: Schema[Foo] = {
+        val foos = list(int)
+          .required[Foo]("foos", _.foos)
+          .addHints(XmlFlattened())
+        struct(foos)(Foo.apply)
+      }
+    }
+    val expected = "foos.1=1&foos.2=2&foos.3=3"
+    checkContent(expected, Foo(List(1, 2, 3)))
+  }
+
+  test("list: flattened custom names") {
+    case class Foo(foos: List[Int])
+    object Foo {
+      implicit val schema: Schema[Foo] = {
+        val foos = list(int)
+          .required[Foo]("foos", _.foos)
+          .addHints(XmlFlattened(), XmlName("x"))
+        struct(foos)(Foo.apply)
+      }
+    }
+    val expected = "x.1=1&x.2=2&x.3=3"
+
+    checkContent(expected, Foo(List(1, 2, 3)))
+  }
+
+  test("recursion") {
+    ignore("Not ready yet")
```

Comment:
the implementation of `lazily` looks good to me. I think we need to amend the datatype of test to carry more information, because the protocol is lossy and the current test wouldn't show much. 

If you had a required field, it will make it naturally easier to test :  

```scala 
case class  IntList(head: Int, tail: Option[IntList]) 

val input = IntList(1, Some(IntList(2, None)))
expected = "head=1&tail.head=2" 
```